akka-cluster/src/main/scala/akka/cluster/JoinConfigCompatChecker.scala

```diff
@@ -0,0 +1,166 @@
+package akka.cluster
+
+import java.util
+
+import akka.actor.{ ActorSystem, ExtendedActorSystem }
+import akka.annotation.InternalApi
+import com.typesafe.config.{ Config, ConfigFactory, ConfigUtil, ConfigValue }
+
+import scala.collection.JavaConverters._
+import scala.collection.{ immutable ⇒ im }
+
+trait JoinConfigCompatChecker {
+
+  /** A Set of configuration keys that are required for this checker */
+  def requiredKeys: Set[String]
+
+  /**
+   * Runs the Config check.
+   *
+   * Implementers are free to define what makes Config entry compatible or not.
+   * We do provide some pre-build checks tough: [[JoinConfigCompatChecker.exists()]] and [[JoinConfigCompatChecker.fullMatch()]]
+   *
+   * @param toCheck - the Config instance to be checked
+   * @param actualConfig - the Config instance containing the actual values
+   * @return a [[ConfigValidation]]. Can be [[Valid]] or [[Invalid]], the later must contain a descriptive list of error messages.
+   */
+  def check(toCheck: Config, actualConfig: Config): ConfigValidation
+}
+
+object JoinConfigCompatChecker {
+
+  /**
+   * Checks that all `requiredKeys` are available in `toCheck` Config.
+   *
+   * @param requiredKeys - a Set of required keys
+   * @param toCheck - the Config instance to be checked
+   */
+  def exists(requiredKeys: Set[String], toCheck: Config): ConfigValidation = {
```

Comment:
internal? otherwise Java API also

Reply:
and this will not even be visible as a static method for Java, when compiled with Scala 2.11, since it's trait + object. Then it would have to be abstract class + object